tensorflow/python/framework/python_op_gen.cc

```diff
@@ -367,6 +432,108 @@ string GenEagerPythonOp::Code() {
+    } else {
+      // Get the dtype of the Tensor
+      const string py_dtype = python_op_gen_internal::DataTypeToPython(arg.type(), "_dtypes.");
+      if (dtype_type.find(py_dtype) != dtype_type.end()) {
+        strings::StrAppend(&type_annotation, "_ops.Tensor[", dtype_type[py_dtype], "]");
+      }
+    }
+
+    type_annotations[arg.name()] = type_annotation;
+  }
+
+  // Mapping output Tensor to its type
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
+    string type_annotation;
+    if (type_annotations.find(arg.type_attr()) != type_annotations.end()) {
+      strings::StrAppend(&type_annotation, "_ops.Tensor[", type_annotations[arg.type_attr()], "]");
+    } else {
+      const string py_dtype = python_op_gen_internal::DataTypeToPython(arg.type(), "_dtypes.");
+      if (dtype_type.find(py_dtype) != dtype_type.end()) {
+        strings::StrAppend(&type_annotation, "_ops.Tensor[", dtype_type[py_dtype], "]");
+      }
+    }
+
+    type_annotations[arg.name()] = type_annotation;
+  }
+
+  return type_annotations;
+}
+
+// Generate TypeVars using attrs
+void GenEagerPythonOp::GenerateTypeVars(std::unordered_map<string, string>& type_annotations) {
+  bool added_typevar = false;
+  for (const auto& attr : op_def_.attr()) {
+    if (attr.type() == "type") {
+      std::vector<string> allowed_types;
+      for (int t : attr.allowed_values().list().type()) {
+        DataType dtype = static_cast<DataType>(t);
+        const string py_dtype = python_op_gen_internal::DataTypeToPython(dtype, "_dtypes.");
+        if (dtype_type.find(py_dtype) != dtype_type.end()) {
+          allowed_types.emplace_back(dtype_type[py_dtype]);
+        }
+      }
+
+      // If all dtypes are allowed, add them all
+      if (allowed_types.empty()) {
+        for (std::pair<string, string> map_dtype : dtype_type) {
+          allowed_types.emplace_back(map_dtype.second);
+        }
+      }
+
+      std::sort(allowed_types.begin(), allowed_types.end());
+
+      string typevar_dtypes;
+      for (std::vector<string>::iterator it = allowed_types.begin(); it != allowed_types.end(); ++it) {
+        if (!typevar_dtypes.empty()) strings::StrAppend(&typevar_dtypes, ", ");
+        strings::StrAppend(&typevar_dtypes, *it);
+      }
+
+      const string type_var_name = type_annotations[attr.name()];
+      strings::StrAppend(&result_, type_var_name, " = TypeVar(\"", type_var_name, "\", ", typevar_dtypes,")\n");
+      added_typevar = true;
+    }
+  }
+
+  if (added_typevar) strings::StrAppend(&result_, "\n");
+}
+
+void GenEagerPythonOp::AddReturnTypeAnnotation(std::unordered_map<string, string>& type_annotations) {
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
+    // Add type annotations to param
+    if (type_annotations.find(arg.name()) != type_annotations.end()) {
+      if (!type_annotations[arg.name()].empty()) {
+        result_.erase(result_.length() - 2);
```

Comment:
Does this remove the ":\n" bit? Add a comment if so. Optional: can we insert our return annotation in the place where that colon is inserted, to avoid erasing (perhaps in AddDefLine)?